safe_transaction_service/events/services/queue_service.py

```diff
@@ -1,276 +1,151 @@
-            exchange=self.exchange_name,
-            exchange_type=ExchangeType.fanout,
-            durable=True,
-            callback=self.on_exchange_declareok,
-        )
-
-    def on_exchange_declareok(self, _unused_frame):
-        """Invoked by pika when RabbitMQ has finished the Exchange.Declare RPC
-        command.
-        Send unsent messages that cannot be sent as due connection errors.
+        self._connections_pool.insert(0, broker_connection)
 
-        :param pika.Frame.Method unused_frame: Exchange.DeclareOk response frame
+    def send_event(self, payload: Dict[str, Any]) -> int:
         """
+        Send an event to rabbitMq exchange
 
-        logger.info("Exchange declared: %s", self.exchange_name)
-        self.send_unsent_events()
-
-
-class SyncQueueService(QueueService):
-    """
-    Synchronous connection with test purpose as we cannot test using gevent connection
-    """
-
-    def __init__(self):
-        super().__init__()
-        self.connect()
-
-    def connect(self) -> BlockingConnection:
+        :param payload: Dict with the payload of the event
         """
-        This method connects to RabbitMq using Blockingconnection.
-        Store in _connection the BlocingConnection object and creates a new channel
+        broker_connection = self.get_connection()
 
-        :return: BlockingConnection
-        """
-        try:
-            self._connection = BlockingConnection(self._connection_parameters)
-            self._channel = self.open_channel()
-            self.setup_exchange()
-            return self._connection
-        except pika.exceptions.AMQPConnectionError:
-            logger.error("Cannot open connection, retrying")
+        event = json.dumps(payload)
+        if broker_connection.publish(event):
+            self.release_connection(broker_connection)
+            return self.send_unsent_events() + 1
+        else:
+            logger.warning("Event can not be sent due any connection error")
+            logger.debug("Adding %s to unsent messages", payload)
+            self.unsent_events.append(event)
 
-    def open_channel(self) -> Channel:
-        """
-        Open a new channel
+        self.release_connection(broker_connection)
+        return 0
 
-        :return: channel opened
+    def send_unsent_events(self) -> int:
         """
-        return self._connection.channel()
+        If connection is ready send the unsent messages list
 
-    def setup_exchange(self):
-        """
-        Setup the exchange on RabbitMQ by invoking the Exchange.Declare RPC
-        command.
+        :return: number of messages sent
         """
-        logger.info("Declaring exchange %s", self.exchange_name)
+        if len(self.unsent_events):
+            broker_connection = self.get_connection()
+            sent_events = 0
+            logger.info("Sending %i not sent messages", len(self.unsent_events))
+            for unsent_message in list(self.unsent_events):
```

Comment:
`list(self.unsent_events)`

This is to prevent race conditions?

Reply:
It was by mistake, I don't know why... but you touch a very interesting topic because two or more threads can call this function and duplicate events, I'll handle it.